test/e2e/workload_api_test.go

```diff
@@ -169,17 +169,68 @@ func TestVSphereUpgradeScaleWorkersUbuntuAPI(t *testing.T) {
 				api.RemoveAllWorkerNodeGroups(), // This gives us a blank slate
 			),
 			vsphere.WithWorkerNodeGroup("worker-0", framework.WithWorkerNodeGroup("worker-0", api.WithCount(1))),
-			vsphere.WithUbuntu123(),
+			vsphere.WithWorkerNodeGroup("worker-1", framework.WithWorkerNodeGroup("worker-1", api.WithCount(1))),
+			vsphere.WithUbuntu124(),
 		),
 	)
 
 	runWorkloadClusterUpgradeFlowAPI(test,
 		api.ClusterToConfigFiller(
-			api.WithCiliumPolicyEnforcementMode(v1alpha1.CiliumPolicyModeAlways),
 			api.WithWorkerNodeGroup("worker-0", api.WithCount(2)),
+			api.RemoveWorkerNodeGroup("worker-1"),
+		),
+		vsphere.WithWorkerNodeGroupConfiguration("worker-1", framework.WithWorkerNodeGroup("worker-2", api.WithCount(1))),
+	)
+}
+
+func TestVSphereUpgradeKubernetesCiliumDisableCSIUbuntuAPI(t *testing.T) {
+	vsphere := framework.NewVSphere(t)
+
+	managementCluster := framework.NewClusterE2ETest(
+		t, vsphere,
+	).WithClusterConfig(
+		api.ClusterToConfigFiller(
+			api.WithControlPlaneCount(1),
+			api.WithWorkerNodeCount(1),
+			api.WithStackedEtcdTopology(),
 		),
 		vsphere.WithUbuntu124(),
 	)
+
+	test := framework.NewMulticlusterE2ETest(t, managementCluster)
+	test.WithWorkloadClusters(
+		framework.NewClusterE2ETest(
+			t, vsphere, framework.WithClusterName(test.NewWorkloadClusterName()),
+		).WithClusterConfig(
+			api.ClusterToConfigFiller(
+				api.WithManagementCluster(managementCluster.ClusterName),
+				api.WithExternalEtcdTopology(1),
+				api.WithControlPlaneCount(1),
+				api.RemoveAllWorkerNodeGroups(), // This gives us a blank slate
+			),
+			vsphere.WithWorkerNodeGroup("worker-0", framework.WithWorkerNodeGroup("worker-0", api.WithCount(1))),
+			vsphere.WithUbuntu123(),
+		),
+	)
+
+	test.CreateManagementCluster()
+	test.RunConcurrentlyInWorkloadClusters(func(wc *framework.WorkloadCluster) {
+		wc.ApplyClusterManifest()
+		wc.WaitForKubeconfig()
+		wc.ValidateClusterState()
+		wc.UpdateClusterConfig(
+			api.ClusterToConfigFiller(
+				api.WithCiliumPolicyEnforcementMode(v1alpha1.CiliumPolicyModeAlways),
+			),
+			api.VSphereToConfigFiller(api.WithDisableCSI(true)),
+			vsphere.WithUbuntu124(),
+		)
+		wc.ApplyClusterManifest()
+		wc.DeleteWorkloadVsphereCSI()
```

Comment:
Do we need to do this manually? I was under the impression that by disabling the CSI the cluster would reconcile and remove these objects automatically.

Reply:
Yes, I believe it has to be done manually.